crates/matrix-sdk-common/src/deserialized_responses.rs

```diff
@@ -7,15 +7,140 @@ use ruma::{
 };
 use serde::{Deserialize, Serialize};
 
-/// The verification state of the device that sent an event to us.
+/// Message shield decoration color abstraction.
+#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
+pub enum ShieldColor {
+    /// Trusted
+    GREEN,
+    /// Important warning
+    RED,
+    /// Low warning
+    GRAY,
+    /// No warning
+    NONE,
+}
+
+/// Directive for clients on how to decorate a decrypted message with this
+/// state.
 #[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ShieldState {
+    /// Recommended indicator color for this message
+    pub color: ShieldColor,
+    /// To display as a "tooltip" for users
+    pub message: Option<String>,
+}
+
+const AUTHENTICITY_NOT_GUARANTEED: &str =
+    "The authenticity of this encrypted message can't be guaranteed on this device.";
+const UNVERIFIED_DEVICE: &str = "Encrypted by an unverified device.";
+const UNSIGNED_DEVICE: &str = "Encrypted by a device not verified by its owner.";
+const UNKNOWN_DEVICE: &str = "Encrypted by an unknown or deleted device.";
+
+impl VerificationState {
+    /// Converts the state to a shield state that can be used by clients
+    pub fn to_shield_state_strict(&self) -> ShieldState {
+        match self {
+            VerificationState::Verified => ShieldState { color: ShieldColor::GREEN, message: None },
+            VerificationState::Unverified(level) => ShieldState {
+                color: ShieldColor::RED,
+                message: match level {
+                    VerificationLevel::UnverifiedIdentity | VerificationLevel::UnsignedDevice => {
+                        Some(UNVERIFIED_DEVICE.to_owned())
+                    }
+                    VerificationLevel::None(link) => match link {
+                        DeviceLinkProblem::MissingDevice => Some(UNKNOWN_DEVICE.to_owned()),
+                        DeviceLinkProblem::InsecureSource => {
+                            Some(AUTHENTICITY_NOT_GUARANTEED.to_owned())
+                        }
+                    },
+                },
+            },
+        }
+    }
```

Comment:
I think we also need another variant (called just `to_shield_state`?) which warns loudly on room keys from insecure sources but doesn't complain about unverified user identities. This should probably be the default for normal rooms (as opposed to high-assurance ones) and is the place where we would be plugging in the TOFU-signed requirement once that's implemented.